File: src/potatoclient/java/MessageProtocol.java
package potatoclient.java;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.io.StringWriter;
import java.io.PrintWriter;
import java.awt.Component;
import static potatoclient.java.Constants.*;

/**
 * Handles all communication protocol with the parent Clojure process.
 * Encapsulates message formatting and JSON serialization.
 */
public class MessageProtocol {
    private final ObjectMapper mapper;
    
    // Pre-allocated message templates to reduce garbage
    private final Map<String, Object> responseTemplate;
    private final Map<String, Object> logTemplate;
    private final Map<String, Object> navTemplate;
    private final Map<String, Object> windowTemplate;
    
    public MessageProtocol(String streamId) {
        this.mapper = new ObjectMapper();
        
        // Initialize templates
        this.responseTemplate = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
        this.logTemplate = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
        this.navTemplate = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
        this.windowTemplate = new ConcurrentHashMap<>(MAP_INITIAL_CAPACITY);
        
        responseTemplate.put("type", "response");
        responseTemplate.put("streamId", streamId);
        
        logTemplate.put("type", "log");
        logTemplate.put("streamId", streamId);
        
        navTemplate.put("type", "navigation");
        navTemplate.put("streamId", streamId);
        
        windowTemplate.put("type", "window");
        windowTemplate.put("streamId", streamId);
    }
    
    public void sendResponse(String status, String data) {
        try {
            Map<String, Object> resp = new HashMap<>(responseTemplate);
            resp.put("status", status);
            resp.put("data", data);
            resp.put("timestamp", System.currentTimeMillis());
            
            sendMessage(resp);
        } catch (Exception e) {
            // Can't use sendException here as it might cause infinite recursion
            e.printStackTrace();
        }
    }
    
    public void sendLog(String level, String message) {
        try {
            Map<String, Object> log = new HashMap<>(logTemplate);
            log.put("level", level);
            log.put("message", message);
            log.put("timestamp", System.currentTimeMillis());
            
            sendMessage(log);
        } catch (Exception e) {
            // Can't use sendException here as it might cause infinite recursion
            e.printStackTrace();
        }
    }
    
    public void sendException(String context, Exception ex) {
        try {
            StringWriter sw = new StringWriter(STRING_WRITER_INITIAL_SIZE);
            PrintWriter pw = new PrintWriter(sw);
            ex.printStackTrace(pw);
            
            Map<String, Object> log = new HashMap<>(logTemplate);
            log.put("level", "ERROR");
            log.put("message", context + ": " + ex.getMessage());
            log.put("stackTrace", sw.toString());
            log.put("timestamp", System.currentTimeMillis());
            
            sendMessage(log);
        } catch (Exception e) {
            // Last resort - print to stderr
            e.printStackTrace();
        }
    }
    
    public void sendNavigationEvent(String eventType, int x, int y, 
                                   Component videoComponent, Map<String, Object> details) {
        try {
            // Get canvas dimensions
            int canvasWidth = videoComponent.getWidth();
            int canvasHeight = videoComponent.getHeight();
            
            // Avoid division by zero
            if (canvasWidth == 0 || canvasHeight == 0) return;
            
            // Convert to NDC coordinates (-1 to 1 range)
            double ndcX = (x / (double)canvasWidth) * 2.0 - 1.0;
            double ndcY = (y / (double)canvasHeight) * 2.0 - 1.0;
            
            Map<String, Object> event = new HashMap<>(MAP_INITIAL_CAPACITY + (details != null ? details.size() : 0));
            event.put("type", eventType);
            event.put("x", x);
            event.put("y", y);
            event.put("ndcX", ndcX);
            event.put("ndcY", ndcY);
            event.put("canvasWidth", canvasWidth);
            event.put("canvasHeight", canvasHeight);
            if (details != null) {
                event.putAll(details);
            }
            
            Map<String, Object> nav = new HashMap<>(navTemplate);
            nav.put("event", event);
            nav.put("timestamp", System.currentTimeMillis());
            
            sendMessage(nav);
        } catch (Exception e) {
            sendException("Navigation event error", e);
        }
    }
    
    public void sendWindowEvent(String eventType, Map<String, Object> details) {
        try {
            Map<String, Object> event = new HashMap<>(2 + (details != null ? details.size() : 0));
            event.put("type", eventType);
            if (details != null) {
                event.putAll(details);
            }
            
            Map<String, Object> windowEvent = new HashMap<>(windowTemplate);
            windowEvent.put("event", event);
            windowEvent.put("timestamp", System.currentTimeMillis());
            
            sendMessage(windowEvent);
        } catch (Exception e) {
            sendException("Window event error", e);
        }
    }
    
    private void sendMessage(Map<String, Object> message) throws Exception {
        String json = mapper.writeValueAsString(message);
        System.out.println(json);
        System.out.flush();
    }
}

File: src/potatoclient/kotlin/MessageProtocol.kt
package potatoclient.kotlin

import com.fasterxml.jackson.databind.ObjectMapper
import java.awt.Component
import java.io.PrintWriter
import java.io.StringWriter
import java.util.concurrent.ConcurrentHashMap

/**
 * Handles all communication protocol with the parent Clojure process.
 * Encapsulates message formatting and JSON serialization.
 */
class MessageProtocol(private val streamId: String) {
    private val mapper = ObjectMapper()
    
    // Pre-allocated message templates to reduce garbage
    private val responseTemplate = ConcurrentHashMap<String, Any>(Constants.MAP_INITIAL_CAPACITY).apply {
        put("type", "response")
        put("streamId", streamId)
    }
    
    private val logTemplate = ConcurrentHashMap<String, Any>(Constants.MAP_INITIAL_CAPACITY).apply {
        put("type", "log")
        put("streamId", streamId)
    }
    
    private val navTemplate = ConcurrentHashMap<String, Any>(Constants.MAP_INITIAL_CAPACITY).apply {
        put("type", "navigation")
        put("streamId", streamId)
    }
    
    private val windowTemplate = ConcurrentHashMap<String, Any>(Constants.MAP_INITIAL_CAPACITY).apply {
        put("type", "window")
        put("streamId", streamId)
    }
    
    fun sendResponse(status: String, data: String) {
        try {
            val resp = HashMap(responseTemplate).apply {
                put("status", status)
                put("data", data)
                put("timestamp", System.currentTimeMillis())
            }
            sendMessage(resp)
        } catch (e: Exception) {
            // Can't use sendException here as it might cause infinite recursion
            System.err.println("[MessageProtocol] Error sending response: ${e.javaClass.name}: ${e.message}")
        }
    }
    
    fun sendLog(level: String, message: String) {
        try {
            val log = HashMap(logTemplate).apply {
                put("level", level)
                put("message", message)
                put("timestamp", System.currentTimeMillis())
            }
            sendMessage(log)
        } catch (e: Exception) {
            // Can't use sendException here as it might cause infinite recursion
            System.err.println("[MessageProtocol] Error sending log: ${e.javaClass.name}: ${e.message}")
        }
    }
    
    fun sendException(context: String, ex: Exception) {
        try {
            val sw = StringWriter(Constants.STRING_WRITER_INITIAL_SIZE)
            val pw = PrintWriter(sw)
            ex.printStackTrace(pw)
            
            val log = HashMap(logTemplate).apply {
                put("level", "ERROR")
                put("message", "$context: ${ex.message}")
                put("stackTrace", sw.toString())
                put("timestamp", System.currentTimeMillis())
            }
            sendMessage(log)
        } catch (e: Exception) {
            // Last resort - print to stderr
            System.err.println("[MessageProtocol] Error sending exception: ${e.javaClass.name}: ${e.message}")
        }
    }
    
    // Pre-allocated constants for event keys
    private companion object {
        const val KEY_TYPE = "type"
        const val KEY_X = "x"
        const val KEY_Y = "y"
        const val KEY_NDC_X = "ndcX"
        const val KEY_NDC_Y = "ndcY"
        const val KEY_CANVAS_WIDTH = "canvasWidth"
        const val KEY_CANVAS_HEIGHT = "canvasHeight"
        const val KEY_EVENT = "event"
        const val KEY_TIMESTAMP = "timestamp"
    }
    
    // Thread-local event map to avoid allocations
    private val threadLocalEventMap = ThreadLocal.withInitial { 
        HashMap<String, Any>(Constants.MAP_INITIAL_CAPACITY * 2)
    }
    
    fun sendNavigationEvent(
        eventType: String,
        x: Int,
        y: Int,
        videoComponent: Component,
        details: Map<String, Any>? = null
    ) {
        try {
            // Get canvas dimensions
            val canvasWidth = videoComponent.width
            val canvasHeight = videoComponent.height
            
            // Avoid division by zero
            if (canvasWidth == 0 || canvasHeight == 0) return
            
            // Pre-calculate reciprocals to avoid division
            val widthReciprocal = 2.0 / canvasWidth
            val heightReciprocal = 2.0 / canvasHeight
            
            // Convert to NDC coordinates (-1 to 1 range)
            val ndcX = x * widthReciprocal - 1.0
            val ndcY = y * heightReciprocal - 1.0
            
            // Reuse thread-local map to avoid allocation
            val event = threadLocalEventMap.get().apply {
                clear()
                put(KEY_TYPE, eventType)
                put(KEY_X, x)
                put(KEY_Y, y)
                put(KEY_NDC_X, ndcX)
                put(KEY_NDC_Y, ndcY)
                put(KEY_CANVAS_WIDTH, canvasWidth)
                put(KEY_CANVAS_HEIGHT, canvasHeight)
                details?.let { putAll(it) }
            }
            
            val nav = HashMap(navTemplate).apply {
                put(KEY_EVENT, HashMap(event)) // Copy event map since it's reused
                put(KEY_TIMESTAMP, System.currentTimeMillis())
            }
            
            sendMessage(nav)
        } catch (e: Exception) {
            sendException("Navigation event error", e)
        }
    }

    fun sendWindowEvent(eventType: String, details: Map<String, Any>? = null) {
        try {
            val event = HashMap<String, Any>(2 + (details?.size ?: 0)).apply {
                put(KEY_TYPE, eventType)
                details?.let { putAll(it) }
            }
            
            val window = HashMap(windowTemplate).apply {
                put(KEY_EVENT, event)
                put(KEY_TIMESTAMP, System.currentTimeMillis())
            }
            
            sendMessage(window)
        } catch (e: Exception) {
            sendException("Window event error", e)
        }
    }
    
    @Suppress("NOTHING_TO_INLINE")
    private inline fun sendMessage(message: Map<String, Any>) {
        val json = mapper.writeValueAsString(message)
        println(json)
        System.out.flush()
    }
}